api/src/main/java/javax/servlet/http/HttpServlet.java

```diff
@@ -367,76 +356,44 @@ protected void doDelete(HttpServletRequest req, HttpServletResponse resp) throws
     protected void doOptions(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
         Method[] methods = getAllDeclaredMethods(this.getClass());
 
-        boolean ALLOW_GET = false;
-        boolean ALLOW_HEAD = false;
-        boolean ALLOW_POST = false;
-        boolean ALLOW_PUT = false;
-        boolean ALLOW_DELETE = false;
-        boolean ALLOW_TRACE = true;
-        boolean ALLOW_OPTIONS = true;
-
-        for (int i = 0; i < methods.length; i++) {
-            String methodName = methods[i].getName();
-
-            if (methodName.equals("doGet")) {
-                ALLOW_GET = true;
-                ALLOW_HEAD = true;
-            } else if (methodName.equals("doPost")) {
-                ALLOW_POST = true;
-            } else if (methodName.equals("doPut")) {
-                ALLOW_PUT = true;
-            } else if (methodName.equals("doDelete")) {
-                ALLOW_DELETE = true;
-            }
-
-        }
-
-        // we know "allow" is not null as ALLOW_OPTIONS = true
+        // we know "allow" is not null OPTIONS method are allowed
         // when this method is invoked
         StringBuilder allow = new StringBuilder();
-        if (ALLOW_GET) {
-            allow.append(METHOD_GET);
-        }
-        if (ALLOW_HEAD) {
-            if (allow.length() > 0) {
-                allow.append(", ");
-            }
-            allow.append(METHOD_HEAD);
-        }
-        if (ALLOW_POST) {
-            if (allow.length() > 0) {
-                allow.append(", ");
-            }
-            allow.append(METHOD_POST);
-        }
-        if (ALLOW_PUT) {
-            if (allow.length() > 0) {
-                allow.append(", ");
-            }
-            allow.append(METHOD_PUT);
-        }
-        if (ALLOW_DELETE) {
-            if (allow.length() > 0) {
-                allow.append(", ");
-            }
-            allow.append(METHOD_DELETE);
-        }
-        if (ALLOW_TRACE) {
-            if (allow.length() > 0) {
-                allow.append(", ");
-            }
-            allow.append(METHOD_TRACE);
-        }
-        if (ALLOW_OPTIONS) {
-            if (allow.length() > 0) {
-                allow.append(", ");
+
+        for (Method value : methods) {
```

Comment:
We should calculate this automatic options response either lazily or when the servlet is constructed.  Calculating it on every option request is inefficient.